### PR TITLE
feat: support CNAME records in customDNS mappings

### PR DESCRIPTION
**Pull request by BenMcH**
This PR is related to #465. Currently, I have only implemented CNAME records, but it would be simple to add additional record types mentioned in #996. 

Please let me know if you'd prefer a different approach, but this PR being merged would allow me to replace pihole with Blocky in my home network. I'm happy to add more record types if they make sense. My understanding of DNS is that most other record types typically have more complex use-cases that may not make sense to implement, as they're usually reserved for serving data such as mail configuration or DNS ownership, neither of which I would expect Blocky to be utilized for, given its local-network scope and risk associated with hosting a publicly accessible DNS server.

Further DNS response types can be added with a simple if statement in both the custom_dns.go and custom_dns_resolver.go.

### CODE REVIEW

**Review by ThinkChaos — COMMENTED** [review]
Thanks for picking this up!

I've got a couple comments about details, but also wanted to bring up maybe we should use [dns.ZoneParser](https://pkg.go.dev/github.com/miekg/dns#ZoneParser) which supports the standard zonefile format instead of making something custom? `UnmarshalYAML` could error if any RR is not an `A`/`AAAA`/`CNAME`.  
That also would have the benefit of allowing to split the mapping into another file via `$INCLUDE`.  
That shouldn't be a blocker for this PR though.

**Comment by kwitsch** [comment]
Agreed that the zone parser looks promising but is out of scope and could be implemented later on.

File: config/custom_dns.go
Comment (BenMcH) [reply config/custom_dns.go]: I'm taking a look at this and don't mind cleaning up while I'm in here :smile: Should input be a string? I was thinking it should be a `map[string]string`. 

I need to run some errands, but I'll check back later and finish cleaning this up then

Comment (BenMcH) [reply config/custom_dns.go]: Nevermind, I see what I'm doing wrong! I was implementing CustomDNSMapping, not Entries.

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go:141]: We need to return this error, can you adjust the function signatures? `processCNAME` will need to change too.

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go:101]: Sorry if it wasn't clear but I mean directly passing the IP without going through a string:
```suggestion
					result := r.processIP(v.A, question)
```

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go:105]:
```suggestion
					result := r.processIP(v.AAAA, question)
```

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go:178]: We'll need to do the same thing for `dns.TypeAAAA` here.

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go:161]: I don't think this error will ever happen, but since we're changing the other functions to return an error, can you  check and return it if it occurs?

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go:94]: Let's add a context cancellation check here. That way if we're in an infinite CNAME loop, we'll have an exit strategy:
```suggestion
	for len(domain) > 0 {
		if err := ctx.Err(); err != nil {
			return nil, err
		}

```

We also need another change to actually trigger it, since ATM we never cancel the context.  
Can you also add a timeout in `Server.resolve`, just above the `switch`?
```go
	ctx, cancel := context.WithTimeout(ctx, 100*s.cfg.Upstreams.Timeout.ToDuration())
	defer cancel()

	switch {
		// ...
```

That way we'll reduce the impact of a bad config, and generally having a high global timeout per request is a good safety measure.

**Review by ThinkChaos — COMMENTED** [review]
A couple new minor things and some follow-up from the previous threads.
Progress is looking great :)

**Comment by BenMcH** [comment]
I think that I've implemented most of the PR feedback. I'm not a regular golang user and have been having a heck of a time implementing the `CustomDNSEntries` unmarshalling. I didn't want to force the rest of the changes to come after that, so I skipped it for now. I'll keep hacking away at it on and off tonight, but if you feel this is ready to ship, I can do a follow-up PR once I figure it out.

**Comment by ThinkChaos** [comment]
If you want to post something that's blocking you feel free and I can take a look :)

I resolved most threads, there's just that and one error in `processIP` left, and then I think you'll get some trouble from the linter.

**Comment by BenMcH** [comment]
> If you want to post something that's blocking you feel free and I can take a look :)

I'm mostly just not sure how to get the nested unmarshalling working. I took the sample posted earlier and tried to run with it. Conceptually, the code makes sense to me, however, when I run it, the UnmarshallYAML function gets called and I'm getting empty strings as the input. I'm sure that I've simply mapped something incorrectly. 

Here's a gist of the error: https://gist.github.com/BenMcH/32137efc5f5d5f0ce9f33bc94ccda796

> I resolved most threads, there's just that and one error in `processIP` left, and then I think you'll get some trouble from the linter.

Sorry for my ignorance, but which error in processIP am I missing? I thought that I had updated all of the calls that could return an error to bubble the error up and out of the function.

I'll take a look at the linter soon! :rocket:

**Comment by ThinkChaos** [comment]
> I'm mostly just not sure how to get the nested unmarshalling working.

The code you posted works for me!  
I see one potential issue: the `fmt.Println("Here", input)` runs before `unmarshal(&input)` so it'll see `input` as an empty string. So maybe it's your debugging that's misleading you?  
And another minor thing, but that shouldn't break anything: `result := make(CustomDNSEntries, len(input))` should be `result := make(CustomDNSEntries, len(parts))`.

> Sorry for my ignorance, but which error in processIP am I missing? I thought that I had updated all of the calls that could return an error to bubble the error up and out of the function.

The one from this thread: https://github.com/0xERR0R/blocky/pull/1352#discussion_r1468227613
`rr, _ := ...` should become `rr, err := ...`, and then do the usual `err != nil` dance.

**Comment by BenMcH** [comment]
Thank you for all of your help on this, @ThinkChaos! Turns out I was overthinking it and failed to assign `*c` to result at the end, which made my DNS queries fail :man_facepalming: 

I'll go ahead and implement the rest of the feedback tonight. 

I think that the processIPs change is implemented. The current diff has the following:

```
rr, err := util.CreateAnswerFromQuestion(question, ip, r.cfg.CustomTTL.SecondsU32())
if err != nil {
	return nil, err
}
```

**Comment by BenMcH** [comment]
Alright, `make lint` is now happy locally. I think the tests should be fine as well, but I could be wrong. When I build from `main` or my branch, reverse DNS is not functional, which could probably be attributed to something in my home network or my blocky config (I don't typically use reverse lookups, so this is the first I've tried them at home).

The reverse DNS and old log file deletion tests are failing on both branches on my machine, so I don't _think_ that I introduced it. If I did introduce them, let me know, and I can continue looking into it.

**Comment by BenMcH** [comment]
Seems I just had to walk away and come back with fresh eyes. I was able to fix the reverse DNS tests and now both `make lint` and `make test` work locally!

**Comment by BenMcH** [comment]
I went ahead and added a variety of tests throughout my branch to validate functionality this PR has added and to hopefully surpass the codecov failures. :crossed_fingers:

**Comment by BenMcH** [comment]
Alright, I put the create answer function on the custom resolver and made a setter func, like I saw in the NextResolver struct and forced an error for the create answer path. My local coverage.html file now shows that the file is 100% covered :grin:

File: docs/configuration.md
Comment (kwitsch) [inline docs/configuration.md:287]: No breaker but just an idea:
If every CNAME that is looped through is added to a slice you could check if it already was in the resolving CNAME chain and break the loop with an error. 🤔

**Review by kwitsch — APPROVED** [review]
Looks good to me.

One small comment how to possibly prevent CNAME loops.

Nice work & communication. 👍

File: docs/configuration.md
Comment (BenMcH) [reply docs/configuration.md]: Great idea! I just pushed up a change to add this to the custom resolver!

Comment (BenMcH) [reply docs/configuration.md]: But it broke the linter :man_facepalming: I'll have that fixed in a few mintues

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go]: Sorry to tack this on last minute, but the logic needs to be adjusted. My bad for not realizing earlier and suggesting the AAAA thing.

Based on [RFC-1034 Section 3.6.2](https://www.rfc-editor.org/rfc/rfc1034#section-3.6.2):
- we don't need to do both A and AAAA, but should use what the question asks for. Looking at what `RewriterResolver.rewriteRequest` does should be a good starting point
- if the question is for a CNAME, then we must not recurse but send the CNAME itself
- a CNAME RR must be the only one for a domain

For the last point I'm not sure what's the best thing to do if the user's config doesn't respect that. Here's a couple options:
- error in `UnmarshalYAML`, the user will notice immediately since they're presumably editing and restarting blocky
- log an error and return a ServFail response. EDE, if enabled, could point the user in the right direction.
- discard the extra RRs and just return the result from the CNAME query

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go:192]: Trying to make what it means obvious even without the context of the code:
```suggestion
	return nil, fmt.Errorf("unsupported customDNS RR type %T", entry)
```

**Review by ThinkChaos — CHANGES_REQUESTED** [review]
Can you delete the `blocky` binary and add it to `.gitignore`?

File: resolver/custom_dns_resolver.go
Comment (BenMcH) [reply resolver/custom_dns_resolver.go]: You're correct. we can remove it from the return signature!

File: resolver/custom_dns_resolver.go
Comment (BenMcH) [reply resolver/custom_dns_resolver.go]: The other changes requested have been pushed. I'll start taking a look at this one now, as it seems like a bigger change than the others

Comment (BenMcH) [reply resolver/custom_dns_resolver.go]: Would it be correct to say that a domain could have multiple CNAME RRs associated but that all RRs should be CNAMEs if any one is a CNAME? 

I believe that it is valid to have 2 CNAMEs for a single domain, pointing to different targets.

Comment (BenMcH) [reply resolver/custom_dns_resolver.go]: Looks like my understanding is wrong. Only a single CNAME value is valid for a given domain. I'll add a check to the UnmarshalYAML function so that users are immediately forced to correct the error rather than blindly ignoring configuration.

Comment (ThinkChaos) [reply resolver/custom_dns_resolver.go]: > Only a single CNAME value is valid for a given domain

Yeah that's my understanding too!

**Comment by BenMcH** [comment]
@ThinkChaos thanks for the thorough review! I think that I've implemented all of your suggestions. We now conditionally resolve A or AAAA records on CNAMEs based on the question type and we prevent blocky from starting if there is more than 1 mapping for a domain and any of them are a CNAME.

**Comment by BenMcH** [comment]
Sorry, I should have run lint locally before pushing. _Now_ it should pass :crossed_fingers: I verified both tests and lint

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go:263]: I think we should do like the following to avoid duplication and be ready for other qtypes we add:
```suggestion
	cnames := resolvedCnames
	cnames = append(cnames, targetWithoutDot)

	// Resolve target recursively
	targetReq := newRequestWithClientID(targetWithoutDot, dns.Type(question.Qtype), clientIP, clientID)

	targetResp, err := r.processRequest(ctx, targetReq, cnames)
	if err != nil {
		return nil, err
	}

	result = append(result, targetResp.Res.Answer...)
```

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go:231]: Can you add an early return here for the case of `question.Qtype == dns.TypeCNAME`?

File: resolver/custom_dns_resolver.go
Comment (ThinkChaos) [inline resolver/custom_dns_resolver.go:24]: Very minor but can you move this with the other data fields below, as opposed to the embedded types here? And do the same in `NewCustomDNSResolver`?

**Review by ThinkChaos — COMMENTED** [review]
Ok last round for me I think :)

**Comment by ThinkChaos** [comment]
IDK why we need to re-approve the CI on every push, I don't remember it being like that before. Maybe because it's your first MR to this repo?

**Comment by BenMcH** [comment]
> IDK why we need to re-approve the CI on every push, I don't remember it being like that before. Maybe because it's your first MR to this repo?

That would make sense. Some github actions are things that you wouldn't want to run unless the code has been reviewed/trusted unless the user is a contributor/trusted. I've been making do with the makefile targets locally though :smile:

**Comment by kwitsch** [comment]
> > IDK why we need to re-approve the CI on every push, I don't remember it being like that before. Maybe because it's your first MR to this repo?
> 
> That would make sense. Some github actions are things that you wouldn't want to run unless the code has been reviewed/trusted unless the user is a contributor/trusted. I've been making do with the makefile targets locally though 😄

As far as I observed the behavior:  
Until your first MR is merged the action runs need reaproval on any change.  
If the first MR is through you are somwhat trusted according to GitHub. 😉

**Comment by 0xERR0R** [comment]
> 
> As far as I observed the behavior: Until your first MR is merged the action runs need reaproval on any change. If the first MR is through you are somwhat trusted according to GitHub. 😉

First time contributors need always approval (this is some kind of protection from GitHub to avoid execution of project workflows on each PR) 🤷

**Comment by BenMcH** [comment]
I think that the e2e test failure may have been a false positive. The errors reported were for an http port timeout and an error communicating with mariadb. I ran e2e tests on my local machine this morning and all tests passed.

**Review by ThinkChaos — APPROVED** [review]

**Comment by hozza** [comment]
Could someone help with #1644? - I can't get simple CNAME mapping working under "customDNS":

```yaml
mappings:
    test.internal: CNAME(host.internal)
    host.internal: 192.168.1.100
```

or even with:

```yaml
mappings:
    test.internal: "CNAME(host.internal)"
    host.internal: 192.168.1.100
```

The error is `Error: unable to load configuration file '/app/config.yml': failed to unmarshal config from /app/config.yml: wrong file structure: invalid custom DNS entry 'CNAME(host.internal)': invalid IP address 'CNAME(host.internal)'`

It seems this code was merged and should be in the latest docker container (?), but I don't know what's wrong. 

I've looked through [the code here](https://github.com/0xERR0R/blocky/pull/1352/files) but can't work out what's wrong. There's also an old discussion that "unresolved" here at #412. The somewhat annoying stale bot closed down @asianpianoman issue at #1644.

**Comment by hozza** [comment]
Maybe this was not merged or deployed properly as it seems like "configuration.md" on the [live site here](https://0xerr0r.github.io/blocky/latest/configuration/#custom-dns) does not match entirely with the docs submitted in this PR.

i.e. the live/latest docs mention you can use CNAME in the customDNS config instead of an IP, but it does not show how as specified on this PR's file configuration.md line 287.

**Comment by hozza** [comment]
It seems I've been on a wild Goose Chase... The feature was removed, and the docs were not updated.

#1973 and file https://github.com/0xERR0R/blocky/pull/1973/files removed the doc parts referencing CNAME's only a few days ago and not published live yet. 

However, I can't find the commit/reasoning why this was removed or see it in the change log though. Can someone point me to the right commit/PR just for closure I've spent awhile on this one lol.

**Pull request merged**
